src/snowflake/cli/plugins/nativeapp/run_processor.py

```diff
@@ -267,89 +267,85 @@ def create_or_upgrade_app(
         with self.use_role(self.app_role):
 
             # 1. Need to use a warehouse to create an application object
-            try:
-                if self.application_warehouse:
-                    self._execute_query(f"use warehouse {self.application_warehouse}")
-            except ProgrammingError as err:
-                generic_sql_error_handler(
-                    err=err, role=self.app_role, warehouse=self.application_warehouse
-                )
+            with self.use_warehouse(self.application_warehouse):
```

Comment:
(This applies to all use_warehouse callsites)
I'm still concerned about the case when the passed in warehouse is `None`, which happens when neither `snowflake.yml` nor the connection configuration gives you a warehouse. What you'll get is the fatal `ClickException` thrown that warehouse cannot be `None`. I don't see a test that covers that.

Also, is that error actionable by the user? By throwing it inside `use_warehouse` we don't have the context of why we need the warehouse that would let us throw a better exception.

I have a proposal here:
1. Make `use_warehouse` take `str`, not `Optional[str]`
2. Refactor `self.application_warehouse` -> `self.use_application_warehouse` and throw the error _there_ if there's no warehouse we can resolve --> the error message becomes a lot more meaningful
3. Same for `self.package_warehouse`
4. Add tests that exercise the new exceptions

This refactor assumes we never need to just get the name of either warehouse. We may need both the existing + `use_` versions.